src/pymor/algorithms/rand_la.py

```diff
@@ -0,0 +1,350 @@
+    iscomplex
+        If `True`, the random vectors are chosen complex.
+
+    Returns
+    -------
+    B
+        |VectorArray| which contains the basis, whose span approximates the range of A.
+    """
+    assert source_product is None or isinstance(source_product, Operator)
+    assert range_product is None or isinstance(range_product, Operator)
+    assert isinstance(A, Operator)
+
+    B = A.range.empty()
+
+    R = A.source.random(num_testvecs, distribution='normal')
+    if iscomplex:
+        R += 1j*A.source.random(num_testvecs, distribution='normal')
+
+    if source_product is None:
+        lambda_min = 1
+    elif lambda_min is None:
+        def mv(v):
+            return source_product.apply(source_product.source.from_numpy(v)).to_numpy()
+
+        def mvinv(v):
+            return source_product.apply_inverse(source_product.range.from_numpy(v)).to_numpy()
+        L = LinearOperator((source_product.source.dim, source_product.range.dim), matvec=mv)
+        Linv = LinearOperator((source_product.range.dim, source_product.source.dim), matvec=mvinv)
+        lambda_min = eigsh(L, sigma=0, which="LM", return_eigenvectors=False, k=1, OPinv=Linv)[0]
+
+    testfail = failure_tolerance / min(A.source.dim, A.range.dim)
+    testlimit = np.sqrt(2. * lambda_min) * erfinv(testfail**(1. / num_testvecs)) * tol
+    maxnorm = np.inf
+    M = A.apply(R)
+
+    while(maxnorm > testlimit):
+        basis_length = len(B)
+        v = A.source.random(distribution='normal')
+        if iscomplex:
+            v += 1j*A.source.random(distribution='normal')
+        B.append(A.apply(v))
+        gram_schmidt(B, range_product, atol=0, rtol=0, offset=basis_length, copy=False)
+        M -= B.lincomb(B.inner(M, range_product).T)
+        maxnorm = np.max(M.norm(range_product))
+
+    return B
+
+
+@defaults('q', 'l')
+def rrf(A, source_product=None, range_product=None, q=2, l=8, return_rand=False, iscomplex=False):
+    """Randomized range approximation of `A`.
+
+    This is an implementation of Algorithm 4.4 in :cite:`HMT11`.
+
+    Given the |Operator| `A`, the return value of this method is the |VectorArray|
+    `Q` whose vectors form an approximate orthonomal basis for the range of `A`.
+
+    Parameters
+    ----------
+    A
+        The |Operator| A.
+    source_product
+        Inner product |Operator| of the source of A.
+    range_product
+        Inner product |Operator| of the range of A.
+    q
+        The number of power iterations.
+    l
+        The block size of the normalized power iterations.
+    return_rand
+        If 'True', the randomly sampled |VectorArray| R is returned. 
+    iscomplex
+        If `True`, the random vectors are chosen complex.
+
+    Returns
+    -------
+    Q
+        |VectorArray| which contains the basis, whose span approximates the range of A.
+    R 
+        The randomly sampled |VectorArray|
+    """
+     assert isinstance(A, Operator)
```

Comment:
There is a space too much here, however, this did not fix the indentation error for me. Something is not right with the code formatting. I noticed that there are quite a lot of trailing spaces that have to be removed. You can check for linting errors locally with `flake8 src/pymor/algorithms/rand_la.py` or configure the pre-commit hook https://docs.pymor.org/latest/developer_docs.html#pre-commit-hooks

Reply:
You'll also see annotations from lintly in the `Files changed` tab of PR.

Reply:
i will fix the error with the code formatting in the next commit. @renefritze i only see the error in line 135 (the unexpected indent), is this the only annotation?

Reply:
Looks, like that's currently the only one, yes.

Reply:
The indentations of the function headers that you have created needs to be fixed.

Reply:
The indentation errors are gone now. However, I still get a lot of linting errors. Most of them are trailing whitespaces and whitespaces/indentations in blank lines.